### PR TITLE
gh-119182: Optimize PyUnicode_FromFormat()

### PR DESCRIPTION
**Pull request by vstinner**
Use strchr() and ucs1lib_find_max_char() to optimize the code path formatting sub-strings between '%' formats.

* Issue: gh-119182

### CODE REVIEW

**Comment by vstinner** [comment]
Results of https://github.com/python/cpython/pull/120248#issuecomment-2160590670 benchmark:

```
$ python3 -m pyperf compare_to ref.json change.json --table
+----------------+--------+----------------------+
| Benchmark      | ref    | change               |
+================+========+======================+
| bench 30       | 219 ns | 202 ns: 1.08x faster |
+----------------+--------+----------------------+
| bench 100      | 262 ns | 216 ns: 1.21x faster |
+----------------+--------+----------------------+
| Geometric mean | (ref)  | 1.09x faster         |
+----------------+--------+----------------------+

Benchmark hidden because not significant (1): bench 3
```

File: Objects/unicodeobject.c [hunk 1]
```diff
@@ -2891,31 +2891,31 @@ unicode_from_format(_PyUnicodeWriter *writer, const char *format, va_list vargs)
...
-            len = p - f;
-
-            if (*p == '\0')
+            else {
+                len = strlen(f);
                 writer->overallocate = 0;
+            }
 
-            if (_PyUnicodeWriter_WriteASCIIString(writer, f, len) < 0)
+            int is_ascii = (ucs1lib_find_max_char((Py_UCS1*)f, (Py_UCS1*)f + len) < 128);
```
Comment (serhiy-storchaka) [inline Objects/unicodeobject.c]: You can run it once, for the whole format string.

Comment (vstinner) [reply Objects/unicodeobject.c]: Right. I modified my PR to run it only once, good idea!

File: Objects/unicodeobject.c [hunk 2]
```diff
@@ -2875,47 +2875,45 @@ unicode_fromformat_arg(_PyUnicodeWriter *writer,
...
 
-    va_list vargs2;
-    const char *f;
-
     // Copy varags to be able to pass a reference to a subfunction.
+    va_list vargs2;
     va_copy(vargs2, vargs);
 
-    for (f = format; *f; ) {
+    int is_ascii = (ucs1lib_find_max_char((Py_UCS1*)format, (Py_UCS1*)format + len) < 128);
```
Comment (serhiy-storchaka) [inline Objects/unicodeobject.c:2888]: Add a comment that this is needed because we use `_PyUnicodeWriter_WriteASCIIString`.

Comment (vstinner) [reply Objects/unicodeobject.c:2888]: Good idea, done.

**Comment by vstinner** [comment]
Merged, thanks for the reviews @serhiy-storchaka.

**Pull request merged**